### PR TITLE
More robust peer banning

### PR DESCRIPTION
**Pull request by quentinlesceller**
Handle the errors and propagate them instead of just logging.

### CODE REVIEW

File: servers/src/grin/seed.rs [hunk 1]
```diff
@@ -157,7 +157,9 @@ fn monitor_peers(
 				let interval = Utc::now().timestamp() - x.last_banned;
 				// Unban peer
 				if interval >= config.ban_window() {
-					peers.unban_peer(x.addr);
+					if let Err(e) = peers.unban_peer(x.addr) {
+						error!("failed to  unban peer {}: {:?}", x.addr, e);
```
Comment (antiochp) [inline servers/src/grin/seed.rs]: Excessive whitespace.

Comment (quentinlesceller) [reply servers/src/grin/seed.rs]: Good catch !

File: p2p/src/peers.rs [hunk 2]
```diff
@@ -605,7 +598,12 @@ impl ChainAdapter for Peers {
 				"Received a bad block {} from  {}, the peer will be banned",
 				hash, peer_info.addr,
 			);
-			self.ban_peer(peer_info.addr, ReasonForBan::BadBlock);
+			self.ban_peer(peer_info.addr, ReasonForBan::BadBlock)
+				.map_err(|e| {
+					let err: chain::Error =
```
Comment (hashmap) [inline p2p/src/peers.rs]: Here and below - perhaps just return error `chain::ErrorKind::Other(format!("ban peer error :{:?}", e)).into()` instaead of defining a temp var and returnning it?

File: p2p/src/peers.rs [hunk 3]
```diff
@@ -253,50 +253,43 @@ impl Peers {
...
-			}
-			Err(e) => error!("Couldn't unban {}: {:?}", peer_addr, e),
-		};
+		// check if peer exist
+		self.get_peer(peer_addr)?;
+		if self.is_banned(peer_addr) {
+			self.update_state(peer_addr, State::Healthy)?;
+			Ok(())
+		} else {
+			error!("Couldn't unban {}: peer is not banned", peer_addr);
```
Comment (hashmap) [inline p2p/src/peers.rs]: I'm not sure we should return error and log it in the same place, assuming we log the error higher on the call stack

File: p2p/src/peers.rs [hunk 4]
```diff
@@ -253,50 +253,43 @@ impl Peers {
...
-					error!("Couldn't unban {}: peer is not banned", peer_addr);
-				}
-			}
-			Err(e) => error!("Couldn't unban {}: {:?}", peer_addr, e),
-		};
+		// check if peer exist
+		self.get_peer(peer_addr)?;
+		if self.is_banned(peer_addr) {
+			self.update_state(peer_addr, State::Healthy)?;
+			Ok(())
```
Comment (hashmap) [inline p2p/src/peers.rs]: I'd prefer to return on the previous line, I remember @antiochp like to be more explicit though

File: p2p/src/peers.rs [hunk 5]
```diff
@@ -253,50 +253,43 @@ impl Peers {
...
+		self.update_state(peer_addr, State::Banned)?;
+
+		match self.get_connected_peer(peer_addr) {
+			Some(peer) => {
+				debug!("Banning peer {}", peer_addr);
+				// setting peer status will get it removed at the next clean_peer
+				peer.send_ban_reason(ban_reason)?;
+				peer.set_banned();
+				peer.stop();
+				let mut peers = match self.peers.try_write_for(LOCK_TIMEOUT) {
```
Comment (hashmap) [inline p2p/src/peers.rs]: Wdyt about writing it like:
```
let mut peers = self.peers.try_write_for(LOCK_TIMEOUT).ok_or_else(|| {
  error!("ban_peer: failed to get peers lock");
  Error::PeerException
})?;
```

File: p2p/src/peers.rs (hunk 2 above)
Comment (quentinlesceller) [reply p2p/src/peers.rs]: I have no chance otherwise Rust complains (cannot infer type thing).

File: p2p/src/peers.rs (hunk 4 above)
Comment (quentinlesceller) [reply p2p/src/peers.rs]: :+1:

File: p2p/src/peers.rs (hunk 3 above)
Comment (quentinlesceller) [reply p2p/src/peers.rs]: indeed

File: p2p/src/peers.rs (hunk 2 above)
Comment (quentinlesceller) [reply p2p/src/peers.rs]: Unfortunately cannot be done. Rust need type annotations for that one.

File: p2p/src/peers.rs (hunk 4 above)
Comment (antiochp) [reply p2p/src/peers.rs]: [totally off-topic]

My concern is mainly around multiple calls in a block of code - 

```
// inconsistent across multiple lines (easy to miss that `three()` returns a Result)
{
    one()?;
    two()?;
    three()
}

// more consistency, less line noise if we introduce additional lines or reorder lines etc.
{
    one()?;
    two()?;
    three()?;
    Ok(())
}
```

Also we definitely have cases where you need the `?` to translate errors appropriately.
Not sure this is an issue here though.

I'm definitely coming around to single line blocks of code returning on that single line as you suggest @hashmap. As long as the errors convert correctly.

```
// Also an option if necessary
{
    Ok(one()?)
}
```

File: p2p/src/peers.rs (hunk 5 above)
Comment (quentinlesceller) [reply p2p/src/peers.rs]: 👍

**Comment by quentinlesceller** [comment]
Will merge once tests pass.

File: p2p/src/peers.rs (hunk 4 above)
Comment (hashmap) [reply p2p/src/peers.rs]: Yeah, but requires more lines, may require to introduce a temp var and last but not least - we unwrap Result on line 3 and then wrap it on line 4, instead directly returning it as is. Same with the last approach. I hope the compiler produces the same code though.

**Pull request merged**